pax/plugins/peak_processing/LocalMinimumClustering.py

```diff
@@ -0,0 +1,131 @@
+                w = baseline_to_subtract - pulse.raw_data.astype(np.float64)
+
+                hits_buffer = np.zeros(2, dtype=datastructure.Hit.get_dtype())
+                adc_to_pe = dsputils.adc_to_pe(self.config, h['channel'])
+                hit_bounds = np.array([[h['left'], x], [x+1, h['right']]], dtype=np.int64)
+                hit_bounds -= pulse.left   # build_hits expects hit bounds relative to pulse start
+                build_hits(w,
+                           hit_bounds=hit_bounds,
+                           hits_buffer=hits_buffer,
+                           adc_to_pe=adc_to_pe,
+                           channel=h['channel'],
+                           noise_sigma_pe=pulse.noise_sigma * adc_to_pe,
+                           dt=self.config['sample_duration'],
+                           start=pulse.left,
+                           pulse_i=pulse_i,
+                           saturation_threshold=self.config['digitizer_reference_baseline'] - pulse.baseline - 0.5)
+
+                new_hits.append(hits_buffer)
+
+            hits = np.concatenate(new_hits)
+
+        # Next, split the peaks, sorting hits to the right peak by their maximum index.
+        # Iterate over left, right bounds of the new peaks
+        boundaries = list(zip([0] + [y+1 for y in split_points], split_points + [float('inf')]))
+        for l, r in boundaries:
+            # Convert to index in event
+            l += peak.left
+            r += peak.left
+
+            # Select hits which have their maximum within this peak bounds
+            # The last new peak must also contain hits at the right bound (though this is unlikely to happen)
+            hs = hits[(hits['index_of_maximum'] >= l) &
+                      (hits['index_of_maximum'] <= r)]
+
+            r = r if r < float('inf') else peak.right
+
+            if not len(hs):
+                raise RuntimeError("Attempt to create a peak without hits in LocalMinimumClustering!")
+
+            yield datastructure.Peak(left=l,
+                                     right=r,
+                                     detector=peak.detector, hits=hs)
+
+
+#  @numba.jit(numba.float64[:])
+# TODO: TESTS!
+def find_split_points(w, min_height, min_ratio):
+    """"Finds local minima in w,
+    whose peaks to the left and right both satisfy:
+      - larger than minimum + min_height
+      - larger than minimum * min_ratio
+    """
+    last_max = 0
+    min_since_max = float('inf')
+    min_since_max_i = 0
+
+    for i, x in enumerate(w):
+        if x < min_since_max:
+            # New minimum since last max
+            min_since_max = x
+            min_since_max_i = i
+
+        if min(last_max, x) > max(min_since_max + min_height,
+                                  min_since_max * min_ratio):
+            # Significant local minimum: tell caller, reset both max and min finder
+            yield min_since_max_i
```

Comment:
Yes, I think it makes sense to try at least. For S2 splitting it should work well, for S1 splitting natural breaks should take care of most of the work. Once we know roughly what our S2 widths will be we can find out what filter settings make sense.